bioio/writers/ome_zarr_writer.py

```diff
@@ -1,431 +1,478 @@
 import math
-from typing import Dict, List, Optional, Tuple
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 
-import bioio_base as biob
+import numpy as np
 import zarr
-from ome_zarr.io import parse_url
-from ome_zarr.scale import Scaler
-from ome_zarr.writer import write_image
-from zarr.storage import default_compressor
+from zarr.codecs import BloscCodec, BloscShuffle
 
-from ..ome_utils import generate_ome_channel_id
 
+def downsample_data(
+    data: np.ndarray, factors: Tuple[int, int, int, int, int]
+) -> np.ndarray:
+    """
+    Downsample a 5D array (T, C, Z, Y, X) by integer factors for Z, Y, X.
 
-class OmeZarrWriter:
-    def __init__(self, uri: biob.types.PathLike):
+    Parameters
+    ----------
+    data : np.ndarray
+        The input 5D array (T, C, Z, Y, X).
+    factors : Tuple[int, int, int, int, int]
+        Downsampling factors (t_factor, c_factor ignored, zf, yf, xf).
+
+    Returns
+    -------
+    np.ndarray
+        Downsampled array.
+    """
+    # TODO: There is probably a much more efficent way to do this
+
+    # Only spatial downsampling (time & channel stay intact)
+    zf, yf, xf = factors[2], factors[3], factors[4]
+    T, C, Z, Y, X = data.shape
+
+    # Clamp to avoid factors > dimension
+    zf, yf, xf = min(zf, Z), min(yf, Y), min(xf, X)
+
+    # Compute how many full blocks fit
+    nz, ny, nx = Z // zf, Y // yf, X // xf
+
+    # Trim off remainders so reshape works
+    d = data[..., : nz * zf, : ny * yf, : nx * xf]
+
+    # Reshape into blocks and average
+    # new shape: (T,C,nz,zf,ny,yf,nx,xf)
+    d = d.reshape(T, C, nz, zf, ny, yf, nx, xf)
+    out = d.mean(axis=(3, 5, 7))
+
+    # Round back if integer
+    if not np.issubdtype(data.dtype, np.floating):
+        out = np.rint(out).astype(data.dtype)
+    return out
+
+
+def default_axes(
+    names: List[str], types: List[str], units: List[Optional[str]]
+) -> List[dict]:
+    """
+    Build axes metadata list from provided names, types, and units.
+
+    Parameters
+    ----------
+    names : List[str]
+        Axis names.
+    types : List[str]
+        Axis types.
+    units : List[Optional[str]]
+        Axis units (or None).
+
+    Returns
+    -------
+    List[dict]
+        List of axis metadata dictionaries.
+    """
+
+    # TODO: should this be a constant or a util or something?
+    axes = []
+    for n, t, u in zip(names, types, units):
```

Comment:
Should we validate these are the same size here?

Reply:
zip raises an error if they're not. would validating add some additional functionality here?